File: sdks/node/e2e/run_e2e.mjs
// Node SDK end-to-end proof driver (W4-04).
//
// Spawns crasher.mjs, which inits the real built SDK and throws an uncaught
// error, then asserts through the Crashlens issues API that the crash arrived
// and was grouped into an issue with platform "node". Uses only Node built-ins
// plus the global fetch (Node 18+), so it needs nothing beyond `npm run build`.
//
// Environment (set by the CI e2e job):
//   CRASHLENS_DSN, CRASHLENS_API_BASE (e.g. http://localhost/api),
//   CRASHLENS_TOKEN, CRASHLENS_ORG_ID, CRASHLENS_PROJECT_ID.
//
// Exit code 0 on success, 1 on any failure or timeout.

import { spawn } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { MARKER } from "./marker.mjs";

const EXPECTED_PLATFORM = "node";
const POLL_TIMEOUT_MS = 60_000;

const here = dirname(fileURLToPath(import.meta.url));

function env(name) {
  const value = process.env[name];
  if (!value) {
    throw new Error(`missing required env var ${name}`);
  }
  return value;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function getJson(url, token) {
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) {
    throw new Error(`GET ${url} -> ${res.status}`);
  }
  return res.json();
}

async function runCrasher(dsn) {
  return new Promise((resolve) => {
    const child = spawn(process.execPath, [join(here, "crasher.mjs")], {
      env: { ...process.env, CRASHLENS_DSN: dsn },
      stdio: "inherit",
    });
    // A non-zero exit is EXPECTED (uncaught exception). We only need it to have
    // flushed the event before exiting.
    child.on("close", (code) => resolve(code));
  });
}

async function main() {
  const dsn = env("CRASHLENS_DSN");
  const apiBase = env("CRASHLENS_API_BASE").replace(/\/$/, "");
  const token = env("CRASHLENS_TOKEN");
  const org = env("CRASHLENS_ORG_ID");
  const project = env("CRASHLENS_PROJECT_ID");

  console.log(
    "running the crasher subprocess (an uncaught exception is expected)...",
  );
  const code = await runCrasher(dsn);
  console.log(`crasher exited with code ${code}`);

  const issuesUrl = `${apiBase}/orgs/${org}/projects/${project}/issues`;
  const deadline = Date.now() + POLL_TIMEOUT_MS;
  let found = null;
  while (Date.now() < deadline) {
    try {
      const listing = await getJson(issuesUrl, token);
      found = (listing.issues ?? []).find((i) =>
        (i.title ?? "").includes(MARKER),
      );
    } catch (err) {
      console.log(`  poll error (retrying): ${err.message}`);
    }
    if (found) {
      break;
    }
    await sleep(1000);
  }

  if (!found) {
    console.error(
      "FAIL: no issue carrying the crasher marker appeared within the timeout",
    );
    process.exit(1);
  }

  const detail = await getJson(`${issuesUrl}/${found.id}`, token);
  const platform = detail?.latest_event?.payload?.platform;
  if (platform !== EXPECTED_PLATFORM) {
    console.error(
      `FAIL: issue platform was ${platform}, expected ${EXPECTED_PLATFORM}`,
    );
    process.exit(1);
  }

  console.log(
    `PASS: uncaught "${found.title}" arrived and grouped with platform=${platform} (issue ${found.id})`,
  );
}

main().catch((err) => {
  console.error(`FAIL: ${err.stack || err.message}`);
  process.exit(1);
});

File: sdks/python/e2e/run_e2e.py
"""Python SDK end-to-end proof driver (W4-04).

Spawns crasher.py, which inits the real SDK and raises an uncaught exception,
then asserts through the Crashlens issues API that the crash arrived and was
grouped into an issue with platform ``python``. Uses only the standard library
so it needs nothing beyond the installed SDK (``pip install -e sdks/python``).

Environment (set by the CI e2e job):
  CRASHLENS_DSN        - DSN the crasher inits with.
  CRASHLENS_API_BASE   - e.g. http://localhost/api
  CRASHLENS_TOKEN      - a session token for the account that owns the project.
  CRASHLENS_ORG_ID     - the owning org id.
  CRASHLENS_PROJECT_ID - the project the DSN key belongs to.

Exit code 0 on success, 1 on any failure or timeout.
"""

import json
import os
import subprocess
import sys
import time
import urllib.error
import urllib.request

from crasher import MARKER

EXPECTED_PLATFORM = "python"
POLL_TIMEOUT_S = 60


def _get_json(url: str, token: str) -> dict:
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {token}"})
    with urllib.request.urlopen(req, timeout=10) as resp:  # noqa: S310 - trusted local URL
        return json.loads(resp.read().decode("utf-8"))


def _find_issue(api_base: str, org: str, project: str, token: str) -> dict | None:
    """Return the issue whose title carries the crasher marker, or None yet."""
    listing = _get_json(
        f"{api_base}/orgs/{org}/projects/{project}/issues", token
    )
    for item in listing.get("issues", []):
        if MARKER in item.get("title", ""):
            return item
    return None


def main() -> int:
    dsn = os.environ["CRASHLENS_DSN"]
    api_base = os.environ["CRASHLENS_API_BASE"].rstrip("/")
    token = os.environ["CRASHLENS_TOKEN"]
    org = os.environ["CRASHLENS_ORG_ID"]
    project = os.environ["CRASHLENS_PROJECT_ID"]

    here = os.path.dirname(os.path.abspath(__file__))
    crasher = os.path.join(here, "crasher.py")

    # Run the crasher. A non-zero exit is EXPECTED (uncaught exception); we only
    # need it to have flushed the event before exiting.
    print("running the crasher subprocess (an uncaught exception is expected)...")
    proc = subprocess.run(  # noqa: S603 - fixed argv, trusted script
        [sys.executable, crasher],
        env={**os.environ, "CRASHLENS_DSN": dsn},
        capture_output=True,
        text=True,
        timeout=30,
    )
    print(f"crasher exited with code {proc.returncode}")

    deadline = time.time() + POLL_TIMEOUT_S
    issue = None
    while time.time() < deadline:
        try:
            issue = _find_issue(api_base, org, project, token)
        except (urllib.error.URLError, TimeoutError) as exc:
            print(f"  poll error (retrying): {exc}")
        if issue is not None:
            break
        time.sleep(1)

    if issue is None:
        print("FAIL: no issue carrying the crasher marker appeared within the timeout")
        return 1

    detail = _get_json(
        f"{api_base}/orgs/{org}/projects/{project}/issues/{issue['id']}", token
    )
    platform = (detail.get("latest_event") or {}).get("payload", {}).get("platform")
    if platform != EXPECTED_PLATFORM:
        print(f"FAIL: issue platform was {platform!r}, expected {EXPECTED_PLATFORM!r}")
        return 1

    print(
        f"PASS: uncaught {issue['title']!r} arrived and grouped with "
        f"platform={platform!r} (issue {issue['id']})"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
